py3status/modules/gazelle.py

```diff
@@ -0,0 +1,171 @@
+# -*- coding: utf-8 -*-
+"""
+Show your ratio on your favorite Gazelle-based private tracker.
+
+This module can get various useful infos about you on a private tracker.
+You can display your ratio, upload size, download size.
+The only requirements are that the private tracker is based on the Gazelle framework,
+for instance Apollo or Redacted, and that the sysops didn't block API access somehow.
+This module can save session cookies, and it is advised to avoid too much opened sessions
+that could potentially get you troubles due to a suspect activity.
+
+Configuration parameters:
+    baseurl: Base URL of the tracker (with protocol and no trailing slash)
+        (default 'https://mytracker.stuff')
+    cache_timeout: How often we refresh this module in seconds (default 60)
+    cookiepath: Path of the file containing identification cookie, can be None if
+        you want to create a new session at each load of the module
+        (default '/var/tmp/gazelle_ratio_cookie.txt')
+    format: See placeholders below (default 'Ratio: {ratio}')
+    limit_required: minimum ratio tolerated by tracker (if defined, overrides value given
+        by API, so you are just likely not to fill it) (default None)
+    limit_warning: custom warning ratio (default 1)
+    password: Account password (default '')
+    username: Account username (default '')
+
+Format placeholders:
+    {ratio} account ratio
+    {download} account download amount
+    {upload} account upload amount
+    {messages} number of unread messages
+    {notifs} number of new notifications
+
+Color options:
+    color_ratio_ok: color when ration is bigger than 1 (default color_good)
+    color_ratio_warning: color when ration is between required ratio and 1
+        (available only when tracker has a dynamic required ratio (default color_warning)
+    color_ratio_ko: color when ratio below required ratio (default color_bad)
```

Comment:
Actually that's impossible here, because the thresholds are partly dynamic because given by the tracker response. Actually you can be all static if you use `limit_required` but I'm gonna remove that because I realize it's totally useless to override value given by tracker. So nope, can't do.

Reply:
Well, I can use the threshold mechanic, user would be able to use it it the `format` variable, but won't be able to define it as a nice threshold dict.

Reply:
From my perspective, it sounds like a `threshold` setting that users can configure to make things working correctly (with the site). A nice example easily can show them how to. Otherwise, `gazelle` will always be fully painted in `bad`, `degraded`, or `good` color with no way to turn it off.

With thresholds, users will be able to paint things their way instead of no choice.
* `\?color=ratio Ratio {ratio}`  <---- This is yours. No customization.
* `Ratio [\?color=ratio {ratio}]` 
* ` [\?color=ratio {ratio}]` 

_`New!`_ Individual colors...
* `Ratio [\?color=ratio {ratio}] Download [\?color=download {download}] Upload [\?color=upload {upload}]`

Heck, we even could colorize messages too...
* `\?color=message {message}`
* `\?color=notif {notifs}`

@tobes and I love rainbows. 🌈  We can rainbow your ratio with this.
```
# bad example
format = '\?color=ratio {ratio}
thresholds = [(0.1, '#fcf'), (0.2, '#fff'), (0.3, '#ffz'), (0.4, '#asdasd')]
```
Et cetera.

Reply:
Try `coin_market` with `markets = 5` config. An example of what it could look like.

Reply:
What I said is that here, threshold value depends of the response of the API. Some users will have a minimum ratio of, let's say, 0.1 and some will have a minimum ratio of 0.9. This value is given by the API. So, except if you can write something like this in the config:
```python
thresholds = [
    (0, "bad")
    (ratio_minimum, "degraded")
    (1, "good")
]
```
with   `ratio_minimum` internally calculated at runtime, I'm afraid it won't do it.
Maybe giving abstract variables like this is possible, and in that's case it's great.

Reply:
The `ratio_minimum` value from API shouldn't change? Rarely?
If yes, then for this site, I find out what `ratio_minimum` is, then I can configure the threshold to...
```
thresholds = [
    (0, "bad")
    (0.75, "degraded")
    (1, "good")
]
```

What if users does not want any colors? Just plain white or grey? I think using `color = '#fff'` wouldn't work for your module because it's going to internally use `response['color']`. So that could be an issue too.

I acknowledge your issue and I'm just saying this because that's the usual way to do things w/ colors, formatters, etc even if it's an issue somewhere else. Maybe @tobes can find a solution by feeding `self.py3.thresholds` some dict too.

Reply:
`ratio_minimum` depends of your download amount and of the proportion of torrents you're seeding. So yeah, it can change quickly.

Anyway, it would be unacceptable to manually set a value when the application can get it for you (this is the case not only for py3status of course).

I was thinking that I could use threshold formatted text anyway, but the threshold dict wouldn't be defined directly by the user, the colors would be set with traditional parameters variable. Then the threshold dict would be defined internally on the module.

Reply:
Maybe leave the threshold dict exposed for other placeholders, eg `download`, `upload`, etc... but overwrite the `ratio` or `requiredratio`.... or I think better yet, make up a new `threshold` name like `auto` or `dynamic` for this purpose.

Reply:
I'll test to put internal variables in the dict, maybe that'll work, what do I know...

Reply:
@lasers when you define multiple thresholds, how do you specify the threshold to use in the `format` variable?

Reply:
`format = '\?color=volume Volume: {percent}`
`self.py3.threshold_get_color(data['percent'], 'volume')`

`sysdata.py`
>    format: output format string
        *(default '[\?color=cpu CPU: {cpu_usage}%], '
        '[\?color=mem Mem: {mem_used}/{mem_total} GB ({mem_used_percent}%)]')*
    thresholds: thresholds to use for color changes
        (default [(0, "good"), (40, "degraded"), (75, "bad")])


`netdata.py`
>    format: display format for this module
        *(default '{nic} [\?color=down LAN(Kb): {down}↓ {up}↑]
        [\?color=total T(Mb): {download}↓ {upload}↑ {total}↕]')*
    thresholds: specify color thresholds to use
        *(default {'down': [(0, 'bad'), (30, 'degraded'), (60, 'good')],
        'total': [(0, 'good'), (400, 'degraded'), (700, 'bad')]})*

Reply:
ok, didn't know about `threshold_get_color`.
I feel the headache coming. Honestly that color threshold system is so awfully designed I want to eat my own limbs.